Mathlib/Data/Tree.lean

```diff
@@ -131,48 +121,66 @@ theorem numLeaves_eq_numNodes_succ (x : Tree α) : x.numLeaves = x.numNodes + 1
 #align tree.num_leaves_eq_num_nodes_succ Tree.numLeaves_eq_numNodes_succ
 
 theorem numLeaves_pos (x : Tree α) : 0 < x.numLeaves := by
-  rw [num_leaves_eq_num_nodes_succ]
-  exact x.num_nodes.zero_lt_succ
+  rw [numLeaves_eq_numNodes_succ]
+  exact x.numNodes.zero_lt_succ
 #align tree.num_leaves_pos Tree.numLeaves_pos
 
 theorem height_le_numNodes : ∀ x : Tree α, x.height ≤ x.numNodes
   | nil => le_rfl
   | node _ a b =>
     Nat.succ_le_succ
-      (max_le (trans a.height_le_numNodes <| a.numNodes.le_add_right _)
-        (trans b.height_le_numNodes <| b.numNodes.le_add_left _))
+      (max_le (_root_.trans a.height_le_numNodes <| a.numNodes.le_add_right _)
+        (_root_.trans b.height_le_numNodes <| b.numNodes.le_add_left _))
 #align tree.height_le_num_nodes Tree.height_le_numNodes
 
 /-- The left child of the tree, or `nil` if the tree is `nil` -/
 @[simp]
 def left : Tree α → Tree α
   | nil => nil
-  | node _ l r => l
+  | node _ l _r => l
 #align tree.left Tree.left
 
 /-- The right child of the tree, or `nil` if the tree is `nil` -/
 @[simp]
 def right : Tree α → Tree α
   | nil => nil
-  | node _ l r => r
+  | node _ _l r => r
 #align tree.right Tree.right
 
 -- mathport name: «expr △ »
--- Notation for making a node with `unit` data
+-- Notation for making a node with `Unit` data
 scoped infixr:65 " △ " => Tree.node ()
 
-/-- Recursion on `tree unit`; allows for a better `induction` which does not have to worry
-  about the element of type `α = unit` -/
+-- porting note: workaround for leanprover/lean4#2049
+section recursor_workarounds
+
+/-- A computable version of `Tree.recOn`. Workaround until Lean has native support for this. -/
+def recOnC {α} {motive : Tree α → Sort u} (t : Tree α) (base : motive Tree.nil)
+  (ind : (a : α) → (l : Tree α) → (r : Tree α) → motive l → motive r → motive (Tree.node a l r))
+  : motive t :=
+  match t with
+  | nil => base
+  | Tree.node a l r => ind a l r (recOnC l base ind) (recOnC r base ind)
+
+@[csimp]
+lemma recOn_Unit_eq_recOnC : @Tree.recOn = @Tree.recOnC := by
+  ext α motive t base ind
+  induction t with
+  | nil => rfl
+  | node a l r ihl ihr =>
+    rw [Tree.recOnC, ←ihl, ←ihr]
+
+end recursor_workarounds
+
 @[elab_as_elim]
 def unitRecOn {motive : Tree Unit → Sort _} (t : Tree Unit) (base : motive nil)
     (ind : ∀ x y, motive x → motive y → motive (x △ y)) : motive t :=
-  t.recOn base fun u => u.recOn ind
-#align tree.unit_rec_on Tree.unitRecOn
+    t.recOn base fun _ => ind
+#noalign tree.unit_rec_on
 
-theorem left_node_right_eq_self : ∀ {x : Tree Unit} (hx : x ≠ nil), x.left △ x.right = x
+theorem left_node_right_eq_self : ∀ {x : Tree Unit} (_hx : x ≠ nil), x.left △ x.right = x
   | nil, h => by trivial
-  | a △ b, _ => rfl
+  | node a l r, _ => rfl
```

Comment:
This is unfortunate; are we sure that there isn't a way to keep the syntax in Lean 4?

Reply:
I don't know.